server/src/fstar_connection.ts

```diff
@@ -0,0 +1,512 @@
+	// that is restricted to a response type and error type?
+	//
+	// Maps query-ids to promises that will be resolved with the appropriate
+	// response.
+	private pending_responses: Map<number, {ip: internalPromise<any, any>, is_stream: boolean}>;
+	private fstar: FStar;
+
+	constructor(fstar: FStar) {
+		// F*'s IDE protocol requires that each request have a unique query-id.
+		// We use a monotonic id.
+		this.last_query_id = 0;
+
+		// Queries may be responded to asynchronously, so we keep a running map
+		// of pending responses to handle query responses from F*.
+		this.pending_responses = new Map();
+
+		// TODO(klinvill): Should try to spawn F* from within this constructor
+		// instead.
+		this.fstar = fstar;
+		this.fstar.proc.stdin?.setDefaultEncoding('utf-8');
+
+		// Register message handlers that will resolve the appropriate pending
+		// response promises for each query.
+		//
+		// The bufferedHandler buffers up received input until it finds a valid
+		// message. The wrapped handler will then be called with the parsed
+		// valid message. This handles receiving fragmented messages or multiple
+		// messages over the stream from the F* process.
+		const bufferedHandler = FStarConnection.bufferedMessageHandlerFactory((msg: object) => this.handleResponse(msg));
+		this.fstar.proc.stdout?.on('data', bufferedHandler);
+
+		// F* error messages are just printed out
+		const fstar_proc_name = this.fstar.lax ? 'fstar lax' : 'fstar';
+		this.fstar.proc.stderr?.on('data', (data) => { console.error(`${fstar_proc_name} stderr: ${data}`); });
+	}
+
+	// Attempts to spawn an F* process, using the given configuration and
+	// filePath, and create a connection to it.
+	//
+	// @throws {Error} from `trySpawnFstar`
+	static tryCreateFStarConnection(fstarConfig: FStarConfig, filePath: URI, debug: boolean, lax?: 'lax') : FStarConnection | undefined {
+		const fstar = FStar.trySpawnFstar(fstarConfig, filePath, debug, lax);
+		if (fstar)
+			return new FStarConnection(fstar);
+		else
+			return undefined;
+	}
+
+	// Kills the F* process and closes the connection
+	close() {
+		this.fstar.proc.kill();
+	}
+
+	async restartSolver(debug: boolean) {
+		this.fstar.killZ3SubProcess(debug);
+
+		// Wait for a second for processes to die before restarting the solver
+		await setTimeout(1000);
+		this.restartSolverRequest(debug);
+	}
+
+	// Gets the configuration for the F* process.
+	fstar_config() : FStarConfig {
+		return this.fstar.config;
+	}
+
+	////////////////////////////////////////////////////////////////////////////////////
+	// Methods to send requests to F*
+	////////////////////////////////////////////////////////////////////////////////////
+
+	// Sends a request to the F* process.
+	//
+	// @param expect_response indicates if a response from F* is expected for
+	// this request.
+	private async request(msg: any, debug: boolean, expect_response = true, is_stream = false): Promise<any> {
```

Comment:
```suggestion
	private async request(msg: any, expect_response = true, is_stream = false): Promise<any> {
```

`debug` should probably be a field on `FStarConnection`.  I don't see a reason why we would want to have different debugging settings per request.

Reply:
The reason is that the debugging settings may change between requests since they're set at an editor level, rather than at a connection level. I suppose we could have debug as the field on the connection anyway and modify the `updateConfigurationSettings()` logic on the server to always set the debug field with the new value for each connection for each documentState. I'll go ahead and implement that change.